pure_pursuit/controller.py

```diff
@@ -36,12 +39,41 @@ def calculate_steering(self): # calc the needed steering angle to course correct
         look_ahead_point = self._calculate_look_ahead_point()
 
         alpha = math.atan2(look_ahead_point[1] - self.coordinates[1], look_ahead_point[0] - self.coordinates[0]) - self.orientation # error angle
-        delta = math.atan2(2*self._car_length*math.sin(alpha), ld)
+        delta = math.atan2(2*self._car_length*math.sin(alpha), self._point_distance(look_ahead_point))
         return max(delta, -self._max_steering_angle) if (delta < 0) else min(delta, self._max_steering_angle)
 
-    # private methods
-
     def _calculate_look_ahead_point(self):
         ld = self.velocity * self._kdd
-        # complete logic here
+        
+        near_point_index = self._find_near_point_index()
+
+        swapped_axis_path = np.swapaxes(self.path[max(near_point_index-4, 0):min(near_point_index+4, len(self.path)-1)], 0 , 1)
+        p = np.poly1d(np.polyfit(swapped_axis_path[0], swapped_axis_path[1], self._poly_fit_deg))
+
+        point_index = 0
+        while (self._point_distance(self.path[near_point_index+point_index]) < ld):
```

Comment:
I think there is a better way than using while loop. While loops in python are the slowest. Also there a duplication of code we can export to a method.
No need to change that in this pr though, just wanted to point that out for a future refactor.

Reply:
removed duplicated code and in the process replaced while loop with for loop. this is done in addition to making the code faster to prevent out_of_bounds indexing of the path when looking for target index.